src/team_lib/ucx/xccl_ucx_lib.c

```diff
@@ -90,9 +90,26 @@ xccl_ucx_reduce_init(xccl_coll_op_args_t *coll_args,
 {
     //TODO alg selection for allreduce shoud happen here
     xccl_ucx_collreq_t *req;
+    xccl_status_t status = XCCL_OK;
     xccl_ucx_coll_base_init(coll_args, team, &req);
-    req->start = xccl_ucx_reduce_linear_start;
-    (*request) = (xccl_tl_coll_req_t*)&req->super;
+    if (!coll_args->alg.set_by_user) {
+        /* Automatic algorithm selection - take knomial */
+        req->start = xccl_ucx_reduce_knomial_start;
```

Comment:
do we need a check for np=1 case? knomial will not handle it correctly

Reply:
Good question. Honestly, i think this should be checked somewhere at the upper level. There is 0 sense to even create any collective teams for NP=1 case. Otherwise this kind of check will be necessary in many places: VMC assumes at least 2 ranks, in Hier team the subgrouping assumes at least 2 ranks, so on.

Reply:
Agree, shouldn't be a part of this PR. We can decide what to do in this case later